splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/SpliceBaseOperation.java

```diff
@@ -553,7 +553,7 @@ public ExecRow apply(@Nullable ExecRow execRow) {
     public void openCore() throws StandardException{
         DataSetProcessor dsp = EngineDriver.driver().processorFactory().chooseProcessor(activation, this);
         activation.getLanguageConnectionContext().getStatementContext().registerExpirable(this, Thread.currentThread());
-        if (dsp.getType() == DataSetProcessor.Type.SPARK && !isOlapServer() && !SpliceClient.isClient()) {
+        if (dsp.getType() == DataSetProcessor.Type.SPARK && !(isOlapServer() && activation.isSubStatement()) && !SpliceClient.isClient()) {
```

Comment:
@jyuanca I'm wondering if this line of code needs re-work.  dsp.getType() tells how we're currently executing, but not what the new request uses.  
How about:

> if (activation.datasetProcessorType().isOlap() && !isOlapServer() && !SpliceClient.isClient())

Reply:
@msirek Thanks for your suggestion. I'll give a try.

Reply:
My last suggestion was incorrect, but I'm wondering if this piece is wrong:
!(isOlapServer() && activation.isSubStatement())  

This is the same as:  !isOlapServer() || !activation.isSubStatement()

isSubStatement is false for all regular statements (anything not a trigger), so this will essentially remove the !isOlapServer() check because the 2nd part is always true and we will now be able to start distributed operations from the Olap server.

Reply:
We should be able to do distributed operation from olap server, if olap server parses a SQL and determined it should executed from spark. In this case, the olap server is also a client. It needs to receive results streamed from spark.

Reply:
OK, I didn't realize this is an intentional change, thanks.  Both paths could use spark.  I didn't know there is an advantage of streaming results through a 2nd RemoteQueryClient.